File: dev-jobs-mcp/dev_jobs_mcp/analyzers/salary.py
"""연봉 통계 분석.

각 소스가 채워둔 salary_min/max/currency 를 기반으로 USD 로 정규화한 뒤
중앙값/평균/분포를 계산한다.
"""
from __future__ import annotations
from statistics import median, mean
from ..models import JobPosting

# 환율 (대략적, 실시간이 필요하면 외부 API 호출로 교체)
FX_TO_USD = {
    "USD": 1.0,
    "EUR": 1.08,
    "GBP": 1.27,
    "CAD": 0.73,
    "AUD": 0.66,
    "SGD": 0.74,
    "JPY": 0.0066,
    "KRW": 0.00072,
    "INR": 0.012,
    "CHF": 1.12,
    "SEK": 0.094,
    "PLN": 0.25,
}


def _to_usd_year(amount: float, currency: str, period: str) -> float | None:
    rate = FX_TO_USD.get((currency or "USD").upper())
    if not rate:
        return None
    usd = amount * rate
    p = (period or "YEAR").upper()
    if p == "YEAR":
        # Sanity: 일부 소스가 시간당 임금을 "YEAR" 로 잘못 보내는 케이스
        # (예: RemoteOK 의 Forge Global $21/h). $5,000 미만이면 시간당으로 재해석.
        if 0 < usd < 5000:
            return usd * 40 * 52
        return usd
    if p == "MONTH":
        return usd * 12
    if p == "HOUR":
        return usd * 40 * 52  # 가정: 주 40시간
    if p == "WEEK":
        return usd * 52
    if p == "DAY":
        return usd * 5 * 52
    return usd


def compute_salary_stats(jobs: list[JobPosting]) -> dict:
    """공고 리스트에서 연봉 통계 산출 (USD 연봉 기준)."""
    samples: list[tuple[float, float]] = []  # (min, max)
    for j in jobs:
        if j.salary_min is None and j.salary_max is None:
            continue
        lo = _to_usd_year(j.salary_min or j.salary_max or 0, j.salary_currency, j.salary_period)
        hi = _to_usd_year(j.salary_max or j.salary_min or 0, j.salary_currency, j.salary_period)
        if lo and hi and lo > 0 and hi > 0:
            samples.append((lo, hi))

    if not samples:
        return {
            "sample_size": 0,
            "note": "연봉 정보가 공개된 공고가 없습니다. 대부분의 회사가 description 에서만 언급합니다.",
        }

    mids = [(lo + hi) / 2 for lo, hi in samples]
    lows = [lo for lo, _ in samples]
    highs = [hi for _, hi in samples]

    return {
        "sample_size": len(samples),
        "currency": "USD (연봉 기준 정규화)",
        "median_mid": int(median(mids)),
        "mean_mid": int(mean(mids)),
        "median_low": int(median(lows)),
        "median_high": int(median(highs)),
        "min": int(min(lows)),
        "max": int(max(highs)),
        "total_jobs_analyzed": len(jobs),
        "jobs_with_salary": len(samples),
        "disclosure_rate": round(len(samples) / len(jobs), 2) if jobs else 0,
    }

File: web/lib/salary.ts
import type { JobSalary } from "@/lib/types";

const SYMBOL: Record<string, string> = { USD: "$", GBP: "£", EUR: "€", CAD: "C$", AUD: "A$", SGD: "S$", CHF: "CHF " };

// 원본 통화 우선 표시(£65,600→£66k), 원본 없으면 USD 환산값 폴백. 연봉 k반올림, 시급/월급 접미사.
export function formatSalary(s?: JobSalary | null): string | null {
  if (!s) return null;
  if (s.currency && (s.min || s.max)) {
    const sym = SYMBOL[s.currency] ?? `${s.currency} `;
    const period = (s.period ?? "YEAR").toUpperCase();
    const hourly = period === "HOUR";
    const unit = (n: number) => (hourly ? `${sym}${n}` : `${sym}${Math.round(n / 1000)}k`);
    const suffix = hourly ? "/시간" : period === "MONTH" ? "/월" : "";
    const body = s.min && s.max ? `${unit(s.min)}–${unit(s.max)}` : unit((s.min ?? s.max)!);
    return `${body}${suffix}`;
  }
  const { min_usd, max_usd } = s;
  if (!min_usd && !max_usd) return null;
  const k = (n: number) => `$${Math.round(n / 1000)}k`;
  return min_usd && max_usd ? `${k(min_usd)}–${k(max_usd)}` : k((min_usd ?? max_usd)!);
}

// USD→KRW 환산 환율(근사 상수). 환율은 몇 % 단위로 움직이고 표시는 '약 N억'이라 충분.
// 갱신: 이 값만 바꾸면 전 화면 반영. (후속으로 무료 일일 API 연동 가능 — 유료 불필요.)
export const USD_TO_KRW = 1380;

// 원화 환산용 USD 연봉 범위. min_usd/max_usd 우선, 없으면 USD 원본(min/max). 연봉(YEAR)만 환산.
function usdYearRange(s: JobSalary): [number, number] | null {
  if ((s.period ?? "YEAR").toUpperCase() !== "YEAR") return null; // 시급/월급은 원화 환산 생략
  let lo: number | undefined;
  let hi: number | undefined;
  if (s.min_usd != null || s.max_usd != null) {
    lo = s.min_usd ?? undefined;
    hi = s.max_usd ?? undefined;
  } else if (s.currency === "USD") {
    lo = s.min;
    hi = s.max;
  } else {
    return null; // 비-USD 원본 + USD 환산값 없음 → 환산 불가
  }
  const a = lo ?? hi;
  const b = hi ?? lo;
  if (a == null || b == null) return null;
  return [a, b];
}

// USD 금액 → 원화 단위 문자열(억/만). 1억 이상은 억(10억 미만 1자리), 미만은 만원.
function krwUnit(usd: number): string {
  const won = usd * USD_TO_KRW;
  const eok = won / 1e8;
  if (eok >= 1) return `${eok >= 10 ? Math.round(eok) : Math.round(eok * 10) / 10}억`;
  const man = Math.round(won / 1e4 / 100) * 100; // 100만 단위 반올림
  return `${man.toLocaleString()}만`;
}

// 연봉을 원화 근사 문자열로. 예: $130k–$160k → "약 1.8억~2.2억 원". 환산 불가 시 null.
export function formatSalaryKrw(s?: JobSalary | null): string | null {
  if (!s) return null;
  const r = usdYearRange(s);
  if (!r) return null;
  const [lo, hi] = r;
  const body = lo !== hi ? `${krwUnit(lo)}~${krwUnit(hi)}` : krwUnit(lo);
  return `약 ${body} 원`;
}
